lib/OpenQA/WebAPI/Plugin/Gru.pm

```diff
@@ -59,65 +90,74 @@ sub enqueue {
     my ($self, $task) = (shift, shift);
     my $args = shift // [];
     my $options = shift // {};
+    my $jobs = shift // [];
 
-    $self->schema->resultset('GruTasks')->create(
+    $args = [$args] if ref $args eq 'HASH';
+
+    my $delay = $options->{run_at} && $options->{run_at} > now() ? $options->{run_at} - now() : 0;
+
+    my $gru = $self->schema->resultset('GruTasks')->create(
         {
             taskname => $task,
-            args     => $args,
             priority => $options->{priority} // 0,
-            run_at   => $options->{run_at} // now()});
+            args     => $args,
+            run_at   => $options->{run_at} // now(),
+            jobs     => $jobs,
+        });
+
+    $self->app->minion->enqueue(
+        $task => $args => {priority => $options->{priority} // 0, delay => $delay, notes => {gru_id => $gru->id}});
 }
 
 package OpenQA::WebAPI::Plugin::Gru::Command::gru;
 use Mojo::Base 'Mojolicious::Command';
+use Mojo::Pg;
+use Data::Dumper;
 
 has usage       => "usage: $0 gru [-o]\n";
 has description => 'Run a gru to process jobs - give -o to exit _o_nce everything is done';
 
+sub delete_gru {
+    my ($self, $id) = @_;
+    my $gru = $self->app->db->resultset('GruTasks')->find($id);
+    $gru->delete() if $gru;
+}
+
 sub cmd_list {
     my ($self) = @_;
-
-    my $tasks = $self->app->schema->resultset('GruTasks');
-    while (my $task = $tasks->next) {
-        use Data::Dumper;
-        print $task->taskname . " " . Dumper($task->args) . "\n";
+    my $tasks = $self->app->minion->backend->list_jobs();
+    foreach my $j (@{$tasks->{jobs}}) {
+        print $j->{task} . " " . Dumper($j->{args}) . " result: " . $j->{result} . "\n";
     }
 }
 
-sub run_first {
-    my ($self) = @_;
-
-    my $dtf   = $self->app->schema->storage->datetime_parser;
-    my $where = {run_at => {'<=', $dtf->format_datetime(DBIx::Class::Timestamps::now())}};
-    my $first = $self->app->schema->resultset('GruTasks')
-      ->search($where, {order_by => [{-desc => 'priority'}, 'id'], rows => 1})->first;
-
-    if ($first) {
-        $self->app->log->debug(sprintf("Running Gru task %d(%s)", $first->id, $first->taskname));
-        my $subref = $self->app->gru->{tasks}->{$first->taskname};
-        if ($subref) {
-            eval { &$subref($self->app, $first->args) };
-            if ($@) {
-                print $@ . "\n";
-                return;
-            }
-            $first->delete;
-            return 1;
-        }
+sub execute_job {
+    my ($self, $job) = @_;
+    unless (my $err = $job->_run) {
```

Comment:
The call to `Minion::Job::_run` is private API, and will break in the future.

Reply:
that would just leave me with https://metacpan.org/pod/Minion#foreground as option, and i would avoid to specify each job id (would defeat the purpose of using a framework for job/task runner)

Reply:
And you are sure that you don't want to spawn a separate process for every job?

Reply:
The good thing about performing every job in a separate process is that you can avoid many many common memory leaks.

Reply:
Rationale in https://github.com/os-autoinst/openQA/pull/1620#issuecomment-381146004 :)

Reply:
I'm not opposed to making `Minion::Job::_run` part of the public API. Just need to think about it for a little bit.

Reply:
@kraih do not me wrong, i'm totally for background processes, and i feel same about foreground ones, they should be used for debugging.